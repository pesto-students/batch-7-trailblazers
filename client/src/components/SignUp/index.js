import React,{useState} from "react";
import Avatar from "@material-ui/core/Avatar";
import Button from "@material-ui/core/Button";
import CircularProgress from "@material-ui/core/CircularProgress";
import CssBaseline from "@material-ui/core/CssBaseline";
import TextField from "@material-ui/core/TextField";
import Link from "@material-ui/core/Link";
import Grid from "@material-ui/core/Grid";
import LockOutlinedIcon from "@material-ui/icons/LockOutlined";
import Typography from "@material-ui/core/Typography";
import { makeStyles } from "@material-ui/core/styles";
import Container from "@material-ui/core/Container";
import axios from 'axios';
import {SERVER_URL} from "../../config";
import {Redirect} from 'react-router';
import { useSnackBar } from '../../customHooks';


const useStyles = makeStyles(theme => ({
  "@global": {
    body: {
      backgroundColor: theme.palette.common.white
    }
  },
  paper: {
    marginTop: theme.spacing(8),
    display: "flex",
    flexDirection: "column",
    alignItems: "center"
  },
  avatar: {
    margin: theme.spacing(1),
    backgroundColor: theme.palette.secondary.main
  },
  form: {
    width: "100%", // Fix IE 11 issue.
    marginTop: theme.spacing(3)
  },
  submit: {
    margin: theme.spacing(3, 0, 2)
  },
  errorMessage : {
    "text-align" : "center",
    color : "red",
  }
}));

export default function SignUp() {
  const classes = useStyles();

  const [user,setUser] = useState ({name :"", email:"",password: ""});
  const [errorMessage, setErrorMessage] = useState('');
  const [signUpSuccess, setSignUpSucess] = useState(false);
  const { openSnackBar } = useSnackBar();
  const [isSaving ,setIsSaving] = useState (false);

  const handleEmailChange = evt => {
    setUser({...user,email :evt.target.value});
  };

  const handlePasswordChange = evt => {
    setUser({...user, password:evt.target.value});
  };

  const handleNameChange = evt => {
    setUser({...user, name : evt.target.value});
  };

  const postData = async (e) => {
    e.preventDefault();
    setIsSaving(true);
    try {
      const response = await axios({
        method : 'post',
        url : `${SERVER_URL}/signup`,
        data : {...user},
        headers: { "Content-Type": "application/json" },
      });
      if(response && response.data && response.data.isSuccess) {
        setErrorMessage('');
        setSignUpSucess(true);
        openSnackBar('success', response.data.message);
      }

    } catch (err) {
      if(err.response) {
        openSnackBar('error',err.response.data.message);
      }
    }
    setIsSaving(false);
  };

  return (
    <Container component="main" maxWidth="xs">
      <CssBaseline />
      <div className={classes.paper}>
        <Avatar className={classes.avatar}>
          <LockOutlinedIcon />
        </Avatar>
        <Typography component="h1" variant="h5">
          Sign up
        </Typography>
        <form className={classes.form} noValidate onSubmit={postData}>
          <Grid container spacing={2}>
            <Grid item xs={12} sm={12}>
              <TextField
                autoComplete="fname"
                name="name"
                variant="outlined"
                required
                fullWidth
                id="firstName"
                label="Name"
                autoFocus
                onChange ={handleNameChange}
              />
            </Grid>
            <Grid item xs={12}>
              <TextField
                variant="outlined"
                required
                fullWidth
                id="email"
                label="Email Address"
                name="email"
                autoComplete="email"
                onChange={handleEmailChange}
              />
            </Grid>
            <Grid item xs={12}>
              <TextField
                variant="outlined"
                required
                fullWidth
                name="password"
                label="Password"
                type="password"
                id="password"
                autoComplete="current-password"
                onChange ={handlePasswordChange}
              />
            </Grid>
            <Grid item xs={12}>
              {errorMessage ? <p className= {classes.errorMessage}>{errorMessage}</p> : ''}
            </Grid>
          </Grid>
          {!isSaving && <Button
            type="submit"
            fullWidth
            variant="contained"
            color="primary"
            className={classes.submit}
          >
            Sign Up
          </Button> }
        
          {isSaving && <Button
            type="submit"
            fullWidth
            variant="contained"
            color="primary"
            className={classes.submit}
            disabled
          >
            <CircularProgress size={24} />
          </Button> }
        
          {signUpSuccess && <Redirect to="/login"></Redirect>}
          <Grid container justify="flex-end">
            <Grid item>
              <Link href="/login" variant="body2">
                Already have an account? Sign in
              </Link>
            </Grid>
          </Grid>
        </form>
      </div>
    </Container>
  
  );
}
