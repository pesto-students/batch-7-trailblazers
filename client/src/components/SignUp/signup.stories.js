import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Signup from './index';

storiesOf('Signup', module)
  .add('Demo', () => (
<Signup></Signup>
  ));