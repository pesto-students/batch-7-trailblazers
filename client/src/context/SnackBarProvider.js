import React, { useState } from 'react';
import SnackBarContext from './SnackBarContext';
import SnackBar from '../components/SnackBar';

const SnackBarProvider = ({ children }) => {
  const [open, setOpen] = useState(false);
  const [snackBarProp, setSnackBarProp] = useState({
    message: '',
    type: 'info'
  });

  const openSnackBar = (type = 'success', message = '') => {
    setSnackBarProp({ message, type });
    setOpen(true);
  };
  const closeSnackBar = () => setOpen(false);
  return (
    <SnackBarContext.Provider value={{ openSnackBar, closeSnackBar }}>
      {children}
      <SnackBar open={open} {...snackBarProp} handleClose={closeSnackBar} />
    </SnackBarContext.Provider>
  );
};
export default SnackBarProvider;
