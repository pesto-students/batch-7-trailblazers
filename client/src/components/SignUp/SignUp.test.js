import React from "react";
import { shallow } from "enzyme";
import SignUp from "./index";
import toJson from "enzyme-to-json";
import SnackBarProvider from '../../context/SnackBarProvider';

describe("<Signup />", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <SnackBarProvider>
        <SignUp />
      </SnackBarProvider>);
  });

  it("should render", () => {
    expect(wrapper.exists()).toBe(true);
  });

  it("should call API", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
