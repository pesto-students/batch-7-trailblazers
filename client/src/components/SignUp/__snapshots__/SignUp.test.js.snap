// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup /> should call API 1`] = `
<ContextProvider
  value={
    Object {
      "closeSnackBar": [Function],
      "openSnackBar": [Function],
    }
  }
>
  <SignUp />
  <SnackBar
    handleClose={[Function]}
    message=""
    open={false}
    type="info"
  />
</ContextProvider>
`;
